crates/flux-refineck/src/checker.rs

```diff
@@ -897,56 +931,93 @@ impl<'ck, 'genv, 'tcx, M: Mode> Checker<'ck, 'genv, 'tcx, M> {
         Ok(())
     }
 
+    fn find_self_ty_fn_sig(
+        &self,
+        self_ty: rustc_middle::ty::Ty<'tcx>,
+        span: Span,
+    ) -> Result<PolyFnSig> {
+        let tcx = self.genv.tcx();
+        let predicates = self
+            .genv
+            .predicates_of(self.def_id)
+            .with_span(span)?
+            .instantiate_identity()
+            .predicates;
+
+        for predicate in &predicates {
+            if let Some(proj_clause) = predicate.as_projection_clause()
+                && let def_id = proj_clause.projection_def_id()
+                && let Some(kind) = self.genv.is_fn_output(def_id)
+                && proj_clause
+                    .self_ty()
+                    .skip_binder_ref()
+                    .to_ty()
+                    .to_rustc(tcx)
+                    == self_ty
+            {
+                return Ok(proj_clause.map(|proj_pred| {
+                    let self_ty = proj_pred.self_ty().to_ty();
+                    let tupled_args = proj_pred.projection_ty.args[1].expect_base().to_ty();
+                    let output = proj_pred.term.to_ty();
+                    FnTraitPredicate { kind, self_ty, tupled_args, output }.fndef_sig()
```

Comment:
I don't remember what we do with this `kind`, but this looks wrong. We need the kind of the corresponding trait predicate, but this always returns `FnOnce` (I think it was also returning `FnOnce` before you changed `is_fn_output`)

If I understand correctly, a bound like `F: Fn(i32) -> i32` will produce two clauses `F: Fn<(i32,)>` and `<F as FnOnce>::Output == i32`. This is finding the `<F as FnOnce>::Output == i32` clause, which doesn't contain the kind, for the kind we need to find the corresponding `F: Fn<(i32,)>`. One way to do this would be to call `split_of_fn_trait_clauses` on the predicates and match against the `FnTraitPredicate` directly